src/storage-controller/src/instance.rs

```diff
@@ -0,0 +1,444 @@
+            Ok(()) => info!(%replica_id, "stopped replica task"),
+            Err(error) => warn!(%replica_id, %error, "replica task failed"),
+        }
+    }
+
+    /// Connects to the replica.
+    ///
+    /// The connection is retried forever (with backoff) and this method returns only after
+    /// a connection was successfully established.
+    async fn connect(&self) -> ReplicaClient<T> {
+        let try_connect = |retry: RetryState| {
+            let addrs = &self.config.location.ctl_addrs;
+            let dests = addrs
+                .iter()
+                .map(|addr| (addr.clone(), self.metrics.clone()))
+                .collect();
+            let version = self.config.build_info.semver_version();
+            let client_params = &self.config.grpc_client;
+
+            async move {
+                StorageGrpcClient::connect_partitioned(dests, version, client_params)
+                    .await
+                    .inspect_err(|error| {
+                        let next_backoff = retry.next_backoff.unwrap();
+                        if retry.i >= mz_service::retry::INFO_MIN_RETRIES {
+                            info!(
+                                replica_id = %self.replica_id, %error,
+                                "error connecting to replica, retrying in {:?}",
+                                next_backoff,
+                            );
+                        } else {
+                            debug!(
+                                replica_id = %self.replica_id,
+                                "error connecting to replica, retrying in {:?}",
+                                next_backoff,
+                            );
+                        }
+                    })
+            }
+        };
+
+        Retry::default()
+            .clamp_backoff(Duration::from_secs(1))
+            .retry_async(try_connect)
+            .await
+            .expect("retries forever")
+    }
+
+    /// Runs the message loop.
+    ///
+    /// Returns (with an `Err`) if it encounters an error condition (e.g. the replica disconnects).
+    /// If no error condition is encountered, the task runs until the controller disconnects from
+    /// the command channel, or the task is dropped.
+    async fn run_message_loop(mut self, mut client: ReplicaClient<T>) -> Result<(), anyhow::Error> {
+        loop {
+            select! {
+                // Command from controller to forward to replica.
+                command = self.command_rx.recv() => {
+                    let Some(mut command) = command else {
+                        // Controller is no longer interested in this replica. Shut down.
+                        break;
+                    };
+
+                    self.specialize_command(&mut command);
+                    client.send(command).await?;
+                },
+                // Response from replica to forward to controller.
+                response = client.recv() => {
```

Comment:
Same concern here. This method is not documented as being cancel safe and since it is written as a normal `async` function it is very easy to accidentally make it non cancel-safe later.

Reply:
Hm, this is calling `GenericClient::recv` which requires implementers to be cancel safe:

```
    /// Implementations of this method **must** be [cancellation safe]. That
    /// means that work must not be lost if the future returned by this method
    /// is dropped.
    ///
    /// [cancellation safe]: https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety
    async fn recv(&mut self) -> Result<Option<R>, anyhow::Error>;
```

It's indeed still easy to write a `GenericClient::recv` impl that's not cancel safe, and we should probably go and check them all, but I consider that outside the scope of this PR.

Reply:
This made me realize that the `SequentialHydration`'s `recv` method is actually not cancel safe! Fixed in https://github.com/MaterializeInc/materialize/pull/28816.